src/main/scala/mesosphere/marathon/core/health/impl/HealthCheckActor.scala

```diff
@@ -264,13 +282,393 @@ object HealthCheckActor {
+
   case class InstancesUpdate(version: Timestamp, instances: Seq[Instance])
+}
+
+/**
+  * This is a special flavour of the [[HealthCheckActor]] which utilizes akka-streams (instead of sending itself Tick
+  * messages) to execute Marathon HTTP/S health checks *only*. Mesos and TCP health checks are still handled by the
+  * [[HealthCheckActor]].
+  *
+  * Stream-based implementation profits from:
+  * - automatic back-pressure meaning that we will not overload the same instance with HTTP requests
+  * - akka "superPool" which effectively limits total number of concurrent HTTP requests
+  *
+  * This implementation copy-pastes helper methods from the HealthCheckActor e.g. [[HealthCheckActor.handleHealthResult()]].
+  * I decided against creating a base trait encapsulating these methods since Marathon health checks are supposed to be
+  * deprecated (!) anyway. If you're reading this a year after it was written then we failed to actually remove this
+  * code - sorry about that.
+  *
+  */
+private[health] class MarathonHttpHealthCheckActor(
+    appDef: AppDefinition,
+    appHealthCheckActor: ActorRef,
+    killService: KillService,
+    healthCheck: MarathonHttpHealthCheck,
+    instanceTracker: InstanceTracker,
+    eventBus: EventStream)(implicit mat: Materializer, ec: ExecutionContext) extends Actor with StrictLogging {
+
+  val healthByInstanceId = TrieMap.empty[Instance.Id, Health]
+
+  import context.system
+  import MarathonHttpHealthCheckActor._
+
+  val superPool = Http().superPool[Instance](connectionContext = ConnectionContext.https(disabledSslContext, sslConfig = Some(disabledSslConfig())))
+
+  override def preStart(): Unit = {
+    logger.info(s"Starting health check actor for app ${appDef.id} version ${appDef.version} and healthCheck ${healthCheck}")
+
+    //Start health checking not after the default first health check
+    val startAfter = math.min(healthCheck.interval.toMillis, HealthCheck.DefaultFirstHealthCheckAfter.toMillis).millis
+
+    Source
+      .tick(startAfter, healthCheck.interval, Tick)
+      .map { t =>
+        logger.debug(s"HealthCheck stream for app ${appDef.id} version ${appDef.version} and healthCheck $healthCheck")
+        t
+      }
+      .mapAsync(1) { _ =>
+        instanceTracker.specInstances(appDef.id)
+      }
+      .map { instances =>
+        purgeStatusOfDoneInstances(instances)
+        instances.collect {
+          case instance if instance.runSpecVersion == appDef.version && instance.isRunning =>
+            logger.debug("Making a health check request for {}", instance.instanceId)
+            (instance, healthCheck)
+        }
+      }
+      .mapConcat(identity)
+      .map {
+        case (instance, httpHealthCheck) =>
+          prepareHttpRequest(appDef, instance, httpHealthCheck)
+      }
+      .via(httpWithBypass)
+      .mapAsync(1) {
+        case result if result.version == appDef.version =>
+          handleHealthResult(result)
+      }
+      .map(updateInstanceHealth)
```

Comment:
You didn't 😉 

This probably should be in `Sink.foreach` since it's a side effect. However, we are modifying the state of the actor here so you probably want to send a message to the actor.

...and you actually did not. Just saw that `healthByInstanceId` is a `TrieMap`. Any reason we are not using `ConcurrentHashMap`? If I'm not mistaken the concurrent hash map is faster and `TrieMap` is slower but offers save iteration which we probably don't need.

Overall I have a bad feeling using an actor to keep a concurrent mutable data structure that is change by a stream. What is the motivation to use an actor overall?